analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LockReleasedAllPaths.Monitor.TryCatch.cs

```diff
@@ -0,0 +1,207 @@
+        }
+        catch (Exception ex)
+        {
+            throw;
+        }
+        finally
+        {
+            Monitor.Exit(obj);
+        }
+    }
+
+    public void Method3(bool condition)
+    {
+        Monitor.Enter(obj); // FN
+
+        if (condition)
+        {
+            throw new Exception();
+        }
+
+        Monitor.Exit(obj);
+    }
+
+    public void Method4(string arg)
+    {
+        Monitor.Enter(obj); // FN
+        try
+        {
+            Console.WriteLine(arg.Length);
+        }
+        catch (Exception)
+        {
+            Monitor.Exit(obj);
+            throw;
+        }
+    }
+
+    public void Method5(string arg)
+    {
+        Monitor.Enter(obj); // Compliant
+        try
+        {
+            Console.WriteLine(arg.Length);
+        }
+        catch (NullReferenceException nre)
+        {
+            Monitor.Exit(obj);
+            throw;
+        }
+        Monitor.Exit(obj);
+    }
+
+    public void Method6(string arg)
+    {
+        Monitor.Enter(obj); // FN
+        try
+        {
+            Console.WriteLine(arg.Length);
+        }
+        catch (NullReferenceException nre) when (nre.Message.Contains("Dummy string"))
+        {
+            Monitor.Exit(obj);
+            throw;
+        }
+    }
+
+    public void Method7(string arg)
+    {
+        Monitor.Enter(obj); // FN
+        try
+        {
+            Console.WriteLine(arg.Length);
+        }
+        catch (Exception ex) when (ex is NullReferenceException)
+        {
+            Monitor.Exit(obj);
+            throw;
+        }
+    }
+
+    public void Method8(string arg)
+    {
+        Monitor.Enter(obj); // Compliant
+        try
+        {
+            Console.WriteLine(arg.Length);
+            Monitor.Exit(obj);
+        }
+        catch (Exception)
+        {
+            Monitor.Exit(obj);
+        }
+    }
+
+    public void Method9(string arg)
+    {
+        Monitor.Enter(obj); // Compliant
+        try
+        {
+            Console.WriteLine(arg.Length);
+            Monitor.Exit(obj);
+        }
+        catch
+        {
+            Monitor.Exit(obj);
+        }
+    }
+
+    public void Method10(string arg)
+    {
+        Monitor.Enter(obj); // Compliant
+        try
+        {
+            Console.WriteLine(arg.Length);
+            Monitor.Exit(obj);
+        }
+        catch(NullReferenceException nre)
+        {
+            Monitor.Exit(obj);
+        }
+        catch (Exception)
+        {
+            Monitor.Exit(obj);
+        }
+    }
+
+    public void Method11(string arg)
+    {
+        Monitor.Enter(obj); // FN
+        try
+        {
+            Console.WriteLine(arg.Length);
+            Monitor.Exit(obj);
+        }
+        catch (NullReferenceException nre)
+        {
+        }
+        catch (Exception)
+        {
+            Monitor.Exit(obj);
+        }
+    }
+
+    public void Method12(string arg)
+    {
+        Monitor.Enter(obj); // FN
```

Comment:
Not a FN, lock is always released
```suggestion
    public void Method12(string arg)
    {
        Monitor.Enter(obj);
```

Reply:
Let's add FN counterpart:
```
        Monitor.Enter(obj); // FN

        try
        {
            throw new NotImplementedException();
        }
        catch (InvalidOperationException)
        {
            Monitor.Exit(obj);
        }

```